src/main/java/org/veupathdb/service/eda/ds/plugin/standalonemap/markers/QuantitativeAggregateConfiguration.java

```diff
@@ -89,4 +77,49 @@ public String serializeAverage(double d) {
       return Double.toString(d);
     }
   }
+
+  public static class CategoricalProportionAggregatorFactory implements ContinuousAggregators.ContinuousAggregatorFactory {
+    private final CategoricalAggregationConfig categoricalConfig;
+    private final Supplier<List<String>> vocabSupplier;
+    private final Set<String> distinctDenominatorValues;
+    private final Set<String> numeratorMatchSet;
+    boolean negationMode;
+
+    public CategoricalProportionAggregatorFactory(CategoricalAggregationConfig categoricalConfig,
+                                                  Supplier<List<String>> vocabSupplier) {
+      Set<String> numeratorValues = new HashSet<>(categoricalConfig.getNumeratorValues());
+      Set<String> vocabulary = new HashSet<>(vocabSupplier.get());
+      Set<String> numeratorValuesNegation = Sets.difference(vocabulary, numeratorValues);
+      // Count one of the following depending on which is cheaper:
+      // 1. Total numerator matches.
+      // 2. Everything that doesn't match the numerator.
+      if (numeratorValuesNegation.size() < numeratorValues.size()) {
+        this.negationMode = true;
+        this.numeratorMatchSet = numeratorValuesNegation;
+      } else {
+        this.negationMode = false;
+        this.numeratorMatchSet = numeratorValues;
+      }
+      Set<String> denominatorValues = new HashSet<>(categoricalConfig.getDenominatorValues());
+      this.distinctDenominatorValues = Sets.difference(denominatorValues, numeratorValues);
+      this.categoricalConfig = categoricalConfig;
+      this.vocabSupplier = vocabSupplier;
+    }
+
+    @Override
+    public MarkerAggregator<Double> create(int index, Function<String, Double> valueQuantifier) {
+      return new CategoricalProportionAggregator(numeratorMatchSet,
+          distinctDenominatorValues,
```

Comment:
I see now why you don't have to do the 'else if' above; it's because you already took out the values in the denominator set that live in the numerator set (which is why they're distinct).  But that contract is not clear in the above class.  Maybe add a check above (CategoricalProportionAggregator:30ish) that confirms the intersection is empty.  Or maybe even move the whole "choose negation" logic up there, though don't know if the Aggregator class is used elsewhere- like maybe you need to split it this way for API reasons.  

Seems like the logic for this is spread across these two classes but only by understanding both can you understand what is really going on.

Reply:
Yes, at first I had all of the logic in the constructor of the `CategoricalProportionAggregator` class, but it wasn't performant since we instantiate it once per distinct geohash value.

Maybe if the factory and the class live in the same file and the constructor is private it will be more clear.